consensus/src/pipeline/virtual_processor/processor.rs

```diff
@@ -304,20 +308,41 @@ impl VirtualStateProcessor {
     }
 
     fn maybe_update_pruning_point_and_candidate(self: &Arc<Self>) {
-        // TODO: call this method from resolve_virtual
+        let virtual_sp = self.virtual_state_store.read().get().unwrap().ghostdag_data.selected_parent;
+        if virtual_sp == self.genesis_hash {
+            return;
+        }
+
+        let ghostdag_data = self.ghostdag_store.get_compact_data(virtual_sp).unwrap();
         let pruning_read_guard = self.pruning_store.upgradable_read();
         let current_pp = pruning_read_guard.pruning_point().unwrap();
+        let current_pp_bs = self.ghostdag_store.get_blue_score(current_pp).unwrap();
         let current_pp_candidate = pruning_read_guard.pruning_point_candidate().unwrap();
-        let virtual_gd = self.virtual_state_store.read().get().unwrap().ghostdag_data.to_compact();
         let (new_pruning_point, new_candidate) =
-            self.pruning_manager.next_pruning_point_and_candidate_by_block_hash(virtual_gd, None, current_pp_candidate, current_pp);
+            self.pruning_manager.next_pruning_point_and_candidate_by_block_hash(ghostdag_data, None, current_pp_candidate, current_pp);
 
         if new_pruning_point != current_pp {
+            let mut past_pruning_points_to_add = Vec::new();
+            for current in self.reachability_service.backward_chain_iterator(virtual_sp, current_pp, false) {
+                let current_header = self.headers_store.get_compact_header_data(current).unwrap();
+                if current_header.pruning_point == current_pp || current_header.blue_score < current_pp_bs + self.pruning_depth {
+                    break;
+                }
+
+                if past_pruning_points_to_add.is_empty() || *past_pruning_points_to_add.last().unwrap() != current_header.pruning_point
```

Comment:
Maybe something like `if !past_pruning_points_to_add.last().is_some_and(|h| **h == current_header.pruning_point) {` is more readable (`is_some_and` is from nightly but we can copy it as a util)